src/cpg_flow/workflow.py

```diff
@@ -0,0 +1,684 @@
+        except nx.NetworkXUnfeasible:
+            logging.error("Circular dependencies found between stages")
+            raise
+
+        logging.info(f"Stages in order of execution:\n{stage_names}")
+        stages = [_stages_d[name] for name in stage_names]
+
+        # Round 5: applying workflow options first_stages and last_stages.
+        if first_stages or last_stages:
+            logging.info("Applying workflow/first_stages and workflow/last_stages")
+            self._process_first_last_stages(stages, dag, first_stages, last_stages)
+        elif only_stages:
+            logging.info("Applying workflow/only_stages")
+            self._process_only_stages(stages, dag, only_stages)
+
+        if not (final_set_of_stages := [s.name for s in stages if not s.skipped]):
+            raise WorkflowError("No stages to run")
+
+        logging.info(
+            f"Final list of stages after applying stage configuration options:\n{final_set_of_stages}"
+        )
+
+        required_skipped_stages = [s for s in stages if s.skipped]
+        if required_skipped_stages:
+            logging.info(
+                f'Skipped stages: {", ".join(s.name for s in required_skipped_stages)}'
+            )
+
+        # Round 6: actually adding jobs from the stages.
+        if not self.dry_run:
+            inputs = get_multicohort()  # Would communicate with metamist.
+            for i, stg in enumerate(stages):
+                logging.info("*" * 60)
+                logging.info(f"Stage #{i + 1}: {stg}")
+                # pipeline setup is now done in MultiCohort only
+                # the legacy version (input_datasets) is still supported
+                # that will create a MultiCohort with a single Cohort
+                if isinstance(inputs, MultiCohort):
+                    stg.output_by_target = stg.queue_for_multicohort(inputs)
+                else:
+                    raise WorkflowError(f"Unsupported input type: {inputs}")
+                if errors := self._process_stage_errors(stg.output_by_target):
+                    raise WorkflowError(
+                        f"Stage {stg} failed to queue jobs with errors: "
+                        + "\n".join(errors)
+                    )
+
+                logging.info("")
+
+        else:
+            self.queued_stages = [stg for stg in _stages_d.values() if not stg.skipped]
+            logging.info(f"Queued stages: {self.queued_stages}")
+
+    @staticmethod
+    def _process_stage_errors(
+        output_by_target: dict[str, StageOutput | None],
+    ) -> list[str]:
+        targets_by_error = defaultdict(list)
+        for target, output in output_by_target.items():
+            if output and output.error_msg:
+                targets_by_error[output.error_msg].append(target)
+        return [
+            f'{error}: {", ".join(target_ids)}'
+            for error, target_ids in targets_by_error.items()
+        ]
+
+
+class SequencingGroupStage(Stage[SequencingGroup], ABC):
```

Comment:
Could we move the stage classes out and import them instead to reduce the complexity of this file?